src/watcherEx.ts

```diff
@@ -0,0 +1,238 @@
+  private pubConnection: RedisConnection | RedisClusterConnection;
+  private subConnection: RedisConnection | RedisClusterConnection;
+  private callbackForAddPolicy: any;
+  private callbackForRemovePolicy: any;
+  private callbackForRemoveFilteredPolicy: any;
+  private callbackForSavePolicy: any;
+  private callbackForAddPolicies: any;
+  private callbackForRemovePolicies: any;
+  private channel = 'casbin';
+
+  /**
+   * newWatcherEx creates a watcherEx on the single Redis.
+   * @param options - https://github.com/luin/ioredis/blob/master/API.md#new-redisport-host-options
+   * @example
+   * newWatcherEx('redis://user:password@redis-service.com:6379')
+   * newWatcherEx('//localhost:6379')
+   * newWatcherEx({
+   *   port: 6379, // Redis port
+   *   host: "127.0.0.1", // Redis host
+   *   family: 4, // 4 (IPv4) or 6 (IPv6)
+   *   password: "auth",
+   *   db: 0,
+   *   channel: "casbin"
+   * })
+   */
+  public static async newWatcherEx(
+    options?: WatcherExOptions | string,
+    addPolicyCb?: any,
+    removePolicyCb?: any,
+    removeFilteredPolicyCb?: any,
+    savePolicyCb?: any,
+    addPoliciesCb?: any,
+    removePoliciesCb?: any
+  ): Promise<RedisWatcherEx> {
+    let channel = '';
+    if (typeof options === 'object' && options.channel) {
+      channel = options.channel;
+    }
+
+    const pubConnection = new RedisConnection(options);
+    const subConnection = new RedisConnection(options);
+    return this.init(
+      pubConnection,
+      subConnection,
+      channel,
+      addPolicyCb,
+      removePolicyCb,
+      removeFilteredPolicyCb,
+      savePolicyCb,
+      addPoliciesCb,
+      removePoliciesCb
+    );
+  }
+
+  /**
+   * newWatcherExWithCluster creates a watcherEx on the Redis cluster.
+   * @param nodes - An array of nodes in the cluster, [{ port: number, host: string }]
+   * @param clusterOptions - https://github.com/luin/ioredis/blob/master/API.md#new-clusterstartupnodes-options
+   * @example
+   * newWatcherExWithCluster([{ port: 6380, host: "127.0.0.1"}, { port: 6381, host: "127.0.0.1"})
+   */
+  public static async newWatcherExWithCluster(
+    nodes: ClusterNode[] = [],
+    clusterOptions: WatcherExClusterOptions = {},
+    addPolicyCb?: any,
+    removePolicyCb?: any,
+    removeFilteredPolicyCb?: any,
+    savePolicyCb?: any,
+    addPoliciesCb?: any,
+    removePoliciesCb?: any
+  ): Promise<RedisWatcherEx> {
+    const pubConnection = new RedisClusterConnection(nodes, clusterOptions);
+    const subConnection = new RedisClusterConnection(nodes, clusterOptions);
+    return this.init(
+      pubConnection,
+      subConnection,
+      clusterOptions.channel,
+      addPolicyCb,
+      removePolicyCb,
+      removeFilteredPolicyCb,
+      savePolicyCb,
+      addPoliciesCb,
+      removePoliciesCb
+    );
+  }
+
+  private static async init(
+    pubConnection: RedisConnection | RedisClusterConnection,
+    subConnection: RedisConnection | RedisClusterConnection,
+    channel?: string,
```

Comment:
Maybe one "settings" or "options" parameter that holds all the init configurations. And create an interface for it. Functions with too many input variables are not too easy to read.